kernel/src/table_changes/mod.rs

```
@@ -0,0 +1,157 @@
+///
+/// [`CommitInfo`]: crate::actions::CommitInfo
+#[derive(Debug)]
+pub struct TableChanges {
+    pub log_segment: LogSegment,
+    table_root: Url,
+    end_snapshot: Snapshot,
+    start_version: Version,
+}
+
+impl TableChanges {
+    /// Creates a new [`TableChanges`] instance for the given version range. This function checks
+    /// these two properties:
+    /// - The change data feed table feature must be enabled in both the start or end versions.
+    /// - The schema at the start and end versions are the same.
+    ///
+    /// Note that this does not check that change data feed is enabled for every commit in the
+    /// range. It also does not check that the schema remains the same for the entire range.
+    ///
+    /// # Parameters
+    /// - `table_root`: url pointing at the table root (where `_delta_log` folder is located)
+    /// - `engine`: Implementation of [`Engine`] apis.
+    /// - `start_version`: The start version of the change data feed
+    /// - `end_version`: The end version of the change data feed. If this is none, this defaults to
+    ///   the newest table version.
+    pub fn try_new(
+        table_root: Url,
+        engine: &dyn Engine,
+        start_version: Version,
+        end_version: Option<Version>,
+    ) -> DeltaResult<Self> {
+        let start_snapshot =
+            Snapshot::try_new(table_root.as_url().clone(), engine, Some(start_version))?;
+        let end_snapshot = Snapshot::try_new(table_root.as_url().clone(), engine, end_version)?;
+
+        // Verify CDF is enabled at the beginning and end of the interval
+        let is_cdf_enabled = |snapshot: &Snapshot| {
+            static ENABLE_CDF_FLAG: &str = "delta.enableChangeDataFeed";
+            snapshot
+                .metadata()
+                .configuration
+                .get(ENABLE_CDF_FLAG)
+                .is_some_and(|val| val == "true")
+        };
+        if !is_cdf_enabled(&start_snapshot) {
```

I think we need to check at every point along the way, not just start and end.

---

True, we could leave that till later. I was hoping to do some checking at this stage since we can catch that error earlier. Should I just remove this check?

---

No I think this is okay, as long as we somehow check at each point. I think if it's cheap to error early we should

---

No extra cost since we needed to get both snapshots anyway.

---

can we add comments here for clarity? (e.g. "we check start/end to fail early if needed but at each point yielding CDF batches we do schema compat check" or something like that)

---

added comments addressing enable cdf flag, schema, and protocol.